### PR TITLE
feat: [CDS-82989]: added AsgShiftTraffic step

### PR DESCRIPTION
**Pull request by vitsafronovici**
## Describe your changes
added AsgShiftTraffic step

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/54944)

### CODE REVIEW

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger messagecheck

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger ut7

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepInfo.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  @Getter(onMethod_ = { @ApiModelProperty(hidden = true) }) @ApiModelProperty(hidden = true) String metadata;
+
+  @Builder(builderMethodName = "infoBuilder")
+  public AsgShiftTrafficStepInfo(ParameterField<List<TaskSelectorYaml>> delegateSelectors,
+      ParameterField<Integer> weight, ParameterField<Boolean> downsizeOldAsg, String asgBlueGreenDeployFqn) {
+    super(delegateSelectors, weight, downsizeOldAsg, asgBlueGreenDeployFqn);
+  }
+  @Override
+  public StepType getStepType() {
+    return AsgBlueGreenSwapServiceStep.STEP_TYPE;
```
Comment (lovish1999) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepInfo.java:59]: AsgShiftTrafficStep

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepParameters.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2022 Harness Inc. All rights reserved.
```
Comment (lovish1999) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepParameters.java:2]: :nit: 2023

File: 125-cd-nextgen/src/main/java/io/harness/serializer/CDNGRegistrars.java [hunk 3]
```diff
@@ -1140,6 +1140,18 @@ public class CDNGRegistrars {
                                            .yamlGroup(YamlGroup.builder().group(StepCategory.STEP.name()).build())
                                            .build())
                    .build())
+          .add(YamlSchemaRootClass.builder()
+                   .entityType(EntityType.ASG_SHIFT_TRAFFIC_STEP)
+                   .availableAtProjectLevel(true)
+                   .availableAtOrgLevel(false)
+                   .availableAtAccountLevel(false)
+                   .clazz(AsgBlueGreenSwapServiceStepNode.class)
```
Comment (lovish1999) [inline 125-cd-nextgen/src/main/java/io/harness/serializer/CDNGRegistrars.java:1148]: ShiftTraffic ?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/pipeline/CDStepInfo.java [hunk 4]
```diff
@@ -128,42 +129,42 @@
...
+              FetchInstanceScriptStepInfo.class, ShellScriptProvisionStepInfo.class, UpdateReleaseRepoStepInfo.class,
+              EcsRunTaskStepInfo.class, ElastigroupDeployStepInfo.class, ElastigroupRollbackStepInfo.class,
+              ElastigroupSetupStepInfo.class, TerragruntPlanStepInfo.class, TerragruntApplyStepInfo.class,
+              TerragruntDestroyStepInfo.class, TerragruntRollbackStepInfo.class, AsgCanaryDeployStepInfo.class,
+              AsgCanaryDeleteStepInfo.class, TasSwapRoutesStepInfo.class, TasSwapRollbackStepInfo.class,
+              TasCanaryAppSetupStepInfo.class, TasBGAppSetupStepInfo.class, TasBasicAppSetupStepInfo.class,
+              TasCommandStepInfo.class, ElastigroupBGStageSetupStepInfo.class, ElastigroupSwapRouteStepInfo.class,
+              TasAppResizeStepInfo.class, TasRollbackStepInfo.class, AsgRollingDeployStepInfo.class,
+              AsgRollingRollbackStepInfo.class, AsgBlueGreenDeployStepInfo.class, AsgBlueGreenRollbackStepInfo.class,
+              AsgShiftTrafficStepInfo.class, TasRollingDeployStepInfo.class, TasRollingRollbackStepInfo.class,
```
Comment (lovish1999) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/pipeline/CDStepInfo.java:155]: Maybe we should add it at the end only. We have not maintained order for other asg steps also. So adding them together is not required.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/orchestration/NgStepRegistrar.java [hunk 5]
```diff
@@ -314,6 +315,7 @@ public Map<StepType, Class<? extends Step>> getEngineSteps() {
     engineSteps.put(AsgBlueGreenDeployStep.STEP_TYPE, AsgBlueGreenDeployStep.class);
     engineSteps.put(AsgBlueGreenRollbackStep.STEP_TYPE, AsgBlueGreenRollbackStep.class);
     engineSteps.put(AsgServiceSettingsStep.STEP_TYPE, AsgServiceSettingsStep.class);
+    engineSteps.put(AsgServiceSettingsStep.STEP_TYPE, AsgShiftTrafficStep.class);
```
Comment (lovish1999) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/orchestration/NgStepRegistrar.java:318]: ShiftTraffic ?

**Comment by vitsafronovici** [comment]
trigger uts

**Review by vitsafronovici — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/54944)* status: 24 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepInfo.java` line 59 at r2](https://reviewable.io/reviews/harness/harness-core/54944#gh-1381306691:-NiJXT9h5_-asGdJ8utb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f694d03b4ba060627103d82e20b33c600c804dee/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepInfo.java#L59)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

AsgShiftTrafficStep. ?
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepParameters.java` line 2 at r2](https://reviewable.io/reviews/harness/harness-core/54944#gh-1381307798:-NiJXU7i3IebWMzZSJsM:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f694d03b4ba060627103d82e20b33c600c804dee/125-cd-nextgen/src/main/java/io/harness/cdng/aws/asg/AsgShiftTrafficStepParameters.java#L2)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

:nit: 2023
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/orchestration/NgStepRegistrar.java` line 318 at r2](https://reviewable.io/reviews/harness/harness-core/54944#gh-1381318482:-NiJXV-61OsshRMB3T1b:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f694d03b4ba060627103d82e20b33c600c804dee/125-cd-nextgen/src/main/java/io/harness/cdng/orchestration/NgStepRegistrar.java#L318)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

ShiftTraffic ? 
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/pipeline/CDStepInfo.java` line 155 at r2](https://reviewable.io/reviews/harness/harness-core/54944#gh-1381317897:-NiJXVgu4X9ORN_MUR6H:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f694d03b4ba060627103d82e20b33c600c804dee/125-cd-nextgen/src/main/java/io/harness/cdng/pipeline/CDStepInfo.java#L155)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

Maybe we should add it at the end only. We have not maintained order for other asg steps also. So adding them together is not required. 
</blockquote></details>

Done.
___
*[`125-cd-nextgen/src/main/java/io/harness/serializer/CDNGRegistrars.java` line 1148 at r2](https://reviewable.io/reviews/harness/harness-core/54944#gh-1381315314:-NiJXWdb0pkX-CdtwMni:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f694d03b4ba060627103d82e20b33c600c804dee/125-cd-nextgen/src/main/java/io/harness/serializer/CDNGRegistrars.java#L1148)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

ShiftTraffic ? 
</blockquote></details>

Done.

**Review by lovish1999 — APPROVED** [review]
LGTM

**Comment by vitsafronovici** [comment]
trigger ut6

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by vitsafronovici** [comment]
trigger review

**Pull request merged**